prov/gni/src/gnix_ep.c

```diff
@@ -39,9 +39,352 @@
+	.recv = gnix_ep_trecv,
+	.recvv = gnix_ep_trecvv,
+	.recvmsg = gnix_ep_trecvmsg,
+	.send = gnix_ep_tsend,
+	.sendv = gnix_ep_tsendv,
+	.sendmsg = gnix_ep_tsendmsg,
+	.inject = gnix_ep_tinject,
+	.senddata = fi_no_tagged_senddata,
+	.senddata = gnix_ep_tsenddata,
+	.injectdata = fi_no_tagged_injectdata,
+	.search = gnix_ep_tsearch,
+};
+
+static int gnix_ep_close(fid_t fid)
+{
+	struct gnix_ep *ep;
+
+	ep = container_of(fid, struct gnix_ep, ep_fid.fid);
+	/* TODO: lots more stuff to do here */
+
+	free(ep);
+	return FI_SUCCESS;
+}
+
+static int gnix_ep_bind(fid_t fid, struct fid *bfid, uint64_t flags)
+{
+	int ret = FI_SUCCESS;
+	struct gnix_ep  *ep;
+	struct gnix_av  *av;
+	struct gnix_cq  *cq;
+
+	ep = container_of(fid, struct gnix_ep, ep_fid.fid);
+
+	if (!bfid)
+		return -FI_EINVAL;
+
+	switch (bfid->fclass) {
+	case FI_CLASS_EQ:
+		ret = -FI_ENOSYS;
+		goto err;
+		break;
+	case FI_CLASS_CQ:
+		cq = container_of(bfid, struct gnix_cq, cq_fid.fid);
+		if (ep->domain != cq->domain)
+			return -FI_EINVAL;
+		if (flags & FI_SEND) {
+			ep->send_cq = cq;
+		}
+		if (flags & FI_RECV) {
+			ep->recv_cq = cq;
+		}
+		if (flags & FI_COMPLETION) {
+			ep->no_want_cqes = 1;
+		}
+		break;
+	case FI_CLASS_AV:
+		av = container_of(bfid, struct gnix_av, av_fid.fid);
+		if (ep->domain != av->domain)
+			return -FI_EINVAL;
+		ep->av = av;
+		break;
+	case FI_CLASS_MR:/*TODO: got to figure this one out */
+	case FI_CLASS_CNTR: /* TODO: need to support cntrs someday */
+	default:
+		ret = -FI_ENOSYS;
+		break;
+	}
+
+err:
+        return ret;
+}
+
+
 int gnix_ep_open(struct fid_domain *domain, struct fi_info *info,
 		 struct fid_ep **ep, void *context)
 {
-	/* TODO: implement this puppy */
-	return -FI_ENOSYS;
+	struct gnix_domain *domain_priv;
+	struct gnix_ep *ep_priv;
+
+	if ((domain == NULL) || (info == NULL) || (ep == NULL))
+		return -FI_EINVAL;
+
+	if (info->ep_type != FI_EP_RDM)
+		return -FI_ENOSYS;
+
+	domain_priv = container_of(domain, struct gnix_domain, domain_fid);
+
+	ep_priv = calloc(1, sizeof *ep_priv);
+	if (!ep) {
+		return -FI_ENOMEM;
+	}
+
+	ep_priv->ep_fid.fid.fclass = FI_CLASS_EP;
+	ep_priv->ep_fid.fid.context = context;
+
+	ep_priv->ep_fid.fid.ops = &gnix_ep_fi_ops;
+	ep_priv->ep_fid.ops = &gnix_ep_ops;
+	ep_priv->domain = domain_priv;
+	ep_priv->type = info->ep_type;
+
+	ep_priv->ep_fid.msg = &gnix_ep_msg_ops;
+	ep_priv->ep_fid.rma = &gnix_ep_rma_ops;
+	ep_priv->ep_fid.tagged = &gnix_ep_tagged_ops;
+	ep_priv->ep_fid.atomic = NULL;
+
+	/*
+	 * TODO, initialize vc hash table
+	 */
+	if (ep_priv->type == FI_EP_RDM) {
+		ep_priv->vc_hash_hndl = NULL;
+		ep_priv->ep_fid.cm = NULL;
+	} else {
+		ep_priv->vc = NULL;
+	}
+
+	/*
+	 * TODO: hookup the progress functions
+	 */
+
+	ep_priv->progress_fn = NULL;
+	ep_priv->rx_progress_fn = NULL;
+
+	/*
+ 	 * TODO: hookup a nic to this ep, may add more later if
+ 	 *  fi_tx_context, etc. is invoked on this endpoing
+ 	 */
+
+        *ep = &ep_priv->ep_fid;
+        return 0;
```

Comment:
Why not do the same ret = FI_SUCCESS for initialization and have this return ret?